types/tablet_v2/wlr_tablet_v2_tool.c

```diff
@@ -0,0 +1,515 @@
+	if (!seat) {
+		return NULL;
+	}
+	struct wlr_tablet_v2_tablet_tool *tool =
+		calloc(1, sizeof(struct wlr_tablet_v2_tablet_tool));
+	if (!tool) {
+		return NULL;
+	}
+
+	tool->wlr_tool = wlr_tool;
+	wl_list_init(&tool->clients);
+
+
+	tool->tool_destroy.notify = handle_wlr_tablet_tool_destroy;
+	wl_signal_add(&wlr_tool->events.destroy, &tool->tool_destroy);
+	wl_list_insert(&seat->tools, &tool->link);
+
+	// We need to create a tablet client for all clients on the seat
+	struct wlr_tablet_seat_client_v2 *pos;
+	wl_list_for_each(pos, &seat->clients, seat_link) {
+		// Tell the clients about the new tool
+		add_tablet_tool_client(pos, tool);
+	}
+
+	wl_signal_init(&tool->events.set_cursor);
+
+	return tool;
+}
+
+struct wlr_tablet_tool_client_v2 *tablet_tool_client_from_resource(struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_tablet_tool_v2_interface,
+		&tablet_tool_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+
+/* Actual protocol foo */
+// https://www.geeksforgeeks.org/move-zeroes-end-array/
+static size_t push_zeroes_to_end(uint32_t arr[], size_t n) {
+	size_t count = 0;
+
+	for (size_t i = 0; i < n; i++) {
+		if (arr[i] != 0) {
+			arr[count++] = arr[i];
+		}
+	}
+
+	size_t ret = count;
+
+	while (count < n) {
+		arr[count++] = 0;
+	}
+
+	return ret;
+}
+
+static ssize_t tablet_tool_button_update(struct wlr_tablet_v2_tablet_tool *tool,
+		uint32_t button, enum zwp_tablet_pad_v2_button_state state) {
+	bool found = false;
+	size_t i = 0;
+	for (; i < tool->num_buttons; ++i) {
+		if (tool->pressed_buttons[i] == button) {
+			found = true;
+			break;
+		}
+	}
+
+	if (state == ZWP_TABLET_PAD_V2_BUTTON_STATE_PRESSED && !found &&
+			tool->num_buttons < WLR_TABLET_V2_TOOL_BUTTONS_CAP) {
+		i = tool->num_buttons++;
+		tool->pressed_buttons[i] = button;
+		tool->pressed_serials[i] = -1;
+	} else {
+		i = -1;
+	}
+	if (state == ZWP_TABLET_PAD_V2_BUTTON_STATE_RELEASED && found) {
+		tool->pressed_buttons[i] = 0;
+		tool->pressed_serials[i] = 0;
+		tool->num_buttons = push_zeroes_to_end(tool->pressed_buttons, WLR_TABLET_V2_TOOL_BUTTONS_CAP);
+		tool->num_buttons = push_zeroes_to_end(tool->pressed_serials, WLR_TABLET_V2_TOOL_BUTTONS_CAP);
+	}
+
+	assert(tool->num_buttons <= WLR_TABLET_V2_TOOL_BUTTONS_CAP);
+	return i;
+}
+
+static inline int64_t timespec_to_msec(const struct timespec *a) {
+	return (int64_t)a->tv_sec * 1000 + a->tv_nsec / 1000000;
+}
+
+static void send_tool_frame(void *data) {
+	struct wlr_tablet_tool_client_v2 *tool = data;
+
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	zwp_tablet_tool_v2_send_frame(tool->resource, timespec_to_msec(&now));
+	tool->frame_source = NULL;
+}
+
+static void queue_tool_frame(struct wlr_tablet_tool_client_v2 *tool) {
+	struct wl_display *display = wl_client_get_display(tool->client);
+	struct wl_event_loop *loop = wl_display_get_event_loop(display);
+	if (!tool->frame_source) {
```

Comment:
Could the frame be queued for a different client?

Reply:
I don't quite the question tbh.
The idea here is to queue it for the client struct of the tool (it's stored there as well).
The only situation where two clients could have a need for tablet events is when they change focus, which has to go over `property_out` which forces the frame inline.